tests/variables.tf

```diff
@@ -1,55 +1,63 @@
 variable "AsgMetrics" {
-  default = ["GroupMinSize", "GroupMaxSize", "GroupDesiredCapacity", "GroupInServiceInstances", "GroupPendingInstances", "GroupStandbyInstances", "GroupTerminatingInstances", "GroupTotalInstances"]
-  type    = "list"
+  type        = list(string)
+  description = "(Optional) Autoscaling group metrics"
+  default     = ["GroupMinSize", "GroupMaxSize", "GroupDesiredCapacity", "GroupInServiceInstances", "GroupPendingInstances", "GroupStandbyInstances", "GroupTerminatingInstances", "GroupTotalInstances"]
 }
 
 variable "AsgNotificationTypes" {
-  default = ["autoscaling:EC2_INSTANCE_LAUNCH", "autoscaling:EC2_INSTANCE_LAUNCH_ERROR", "autoscaling:EC2_INSTANCE_TERMINATE", "autoscaling:EC2_INSTANCE_TERMINATE_ERROR", "autoscaling:TEST_NOTIFICATION"]
-  type    = "list"
+  type        = list(string)
+  description = "(Optional) Autoscaling group notification types"
+  default     = ["autoscaling:EC2_INSTANCE_LAUNCH", "autoscaling:EC2_INSTANCE_LAUNCH_ERROR", "autoscaling:EC2_INSTANCE_TERMINATE", "autoscaling:EC2_INSTANCE_TERMINATE_ERROR", "autoscaling:TEST_NOTIFICATION"]
 }
 
 variable "build_id" {
-  default = ""
-  type    = "string"
+  type        = string
+  description = "(Required) Build ID"
```

Comment:
Why is this one being changed to required?

Reply:
In the current form, if you don't provide a build ID, you get an error. 

```
Error: Invalid function argument

  on locals.tf line 3, in locals:
   3:   build_id       = "${substr(element(split(":", local.full_build_id), 1), 0, 8)}${substr(element(split(":", local.full_build_id), 1), 9, 4)}" #extract node portion of uuid (last 6 octets) for brevity
    |----------------
    | local.full_build_id is null

Invalid value for "str" parameter: argument must not be null.


Error: Invalid function argument

  on locals.tf line 3, in locals:
   3:   build_id       = "${substr(element(split(":", local.full_build_id), 1), 0, 8)}${substr(element(split(":", local.full_build_id), 1), 9, 4)}" #extract node portion of uuid (last 6 octets) for brevity
    |----------------
    | local.full_build_id is null

Invalid value for "str" parameter: argument must not be null.
```

Reply:
Also, if it doesn't have a default value, I think of it as "required" since Terraform will ask for the value.

Reply:
> Also, if it doesn't have a default value, I think of it as "required" since Terraform will ask for the value.

That is the definition of required, yes. But it _had_ a default, which is being removed. But there is logic so that if the `build_id` is not provided, then one is auto-generated. By making it required, that logic no longer ever triggers.

Should be able to just change this line in the test, to check for `null`, then `local.full_build_id` will not be null.

```
full_build_id  = "${var.build_id == null ? format("notcb:%s", uuid()) : var.build_id}"
```

Reply:
That should do it.